### DISCUSSION TITLE
DataGridTemplateColumn的Width属性Binding困惑，希望各位老师指点迷津.

### DISCUSSION THREAD

**Discussion by zzu-pawn** [BYJRK/DotNet-Discussions#61] [WPF 开发]
希望DataGrid的模板列动态调整宽度。遇到了两个问题。
我的View中的实现:Width="{Binding ClientNames, Converter={converters:ClientNamesToWidthConverter}}"
1. 调试过程中并没有进入我的转换器
2. 因为DataGrid第一次显示时候，模板列的宽度还无法计算出来，等到后续计算后，模板列宽度并没有生效

**Comment by eijjxx** [comment]
这个我也很头疼。最终我的解决办法是直接继承一个DataGrid，在SizeChanged的时候重新计算DataGrid.Columns里的各个宽度。
直接用Behavior也可以实现，我感觉这个只能在cs里想办法，xaml里没找到太好的办法

**Reply by zzu-pawn** [reply]
谢谢。我投机取巧了。我是这样解决的。在XAML中创建了一个组件（比如TextBlock)。
1.         <TextBlock
            x:Name="ProxyEle"
            DataContext="{Binding}"
            Visibility="Collapsed" />
2.    <DataGridTemplateColumn Width="{Binding Source={x:Reference ProxyEle}, Path=DataContext.XXX>
通过以上两个步骤可以成功获取到ViewModel中的目标属性

**Reply by zzu-pawn** [reply]
DataGridTemplateColumn 不在Visual Tree中，无法通过Relative或者ElementName来定位元素，从而无法获取DataContext。但是可以通过x:Reference 定位到目标控件，获取DataContext.
